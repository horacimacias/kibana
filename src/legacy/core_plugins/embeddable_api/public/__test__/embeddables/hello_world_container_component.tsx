/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { Container, Embeddable } from 'plugins/embeddable_api/index';
import React, { Component, RefObject } from 'react';

import { EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
import { EuiSpacer } from '@elastic/eui';

interface Props {
  container: Container;
}

interface State {
  embeddables: { [key: string]: Embeddable };
  loaded: { [key: string]: boolean };
}

export class HelloWorldContainerComponent extends Component<Props, State> {
  private roots: { [key: string]: RefObject<HTMLDivElement> } = {};
  private mounted: boolean = false;
  private inputSubscription?: Subscription;
  private outputSubscription?: Subscription;

  public constructor(props: Props) {
    super(props);

    Object.values(this.props.container.getInput().panels).forEach(panelState => {
      this.roots[panelState.embeddableId] = React.createRef();
    });
    this.state = {
      loaded: {},
      embeddables: {},
    };
  }

  public async componentDidMount() {
    this.mounted = true;

    Object.values(this.props.container.getInput().panels).forEach(panelState => {
      this.renderEmbeddable(panelState.embeddableId);
    });

    this.inputSubscription = this.props.container.getInput$().subscribe(() => {
      Object.values(this.props.container.getInput().panels).forEach(async panelState => {
        if (this.roots[panelState.embeddableId] === undefined) {
          this.roots[panelState.embeddableId] = React.createRef();
        }

        if (this.state.embeddables[panelState.embeddableId] === undefined) {
          const embeddable = await this.props.container.getEmbeddable(panelState.embeddableId);
          const node = this.roots[panelState.embeddableId].current;
          if (this.mounted && node !== null && embeddable) {
            embeddable.renderInPanel(node, this.props.container);

            this.setState(prevState => ({
              loaded: { ...prevState.loaded, [panelState.embeddableId]: true },
            }));
          }
        }
      });
    });

    this.outputSubscription = this.props.container.getOutput$().subscribe(() => {
      const embeddablesLoaded = this.props.container.getOutput().embeddableLoaded;
      Object.keys(embeddablesLoaded).forEach(async id => {
        const loaded = embeddablesLoaded[id];
        if (loaded && !this.state.loaded[id]) {
          const embeddable = await this.props.container.getEmbeddable(id);
          const node = this.roots[id].current;
          if (this.mounted && node !== null && embeddable) {
            embeddable.renderInPanel(node, this.props.container);
            this.setState({ loaded: embeddablesLoaded });
          }
        }
      });
    });
  }

  public componentWillUnmount() {
    this.props.container.destroy();

    if (this.inputSubscription) {
      this.inputSubscription.unsubscribe();
    }
    if (this.outputSubscription) {
      this.outputSubscription.unsubscribe();
    }
  }

  public renderList() {
    const list = Object.values(this.props.container.getInput().panels).map(panelState => {
      const item = (
        <EuiFlexItem key={panelState.embeddableId}>
          <div style={{ height: '400px' }} ref={this.roots[panelState.embeddableId]} />
        </EuiFlexItem>
      );
      return item;
    });
    return list;
  }

  public render() {
    return (
      <div>
        <h2>HELLO WORLD! These are my precious embeddable children:</h2>
        <EuiSpacer size="l" />
        <EuiFlexGroup>{this.renderList()}</EuiFlexGroup>
      </div>
    );
  }

  private async renderEmbeddable(id: string) {
    if (this.state.embeddables[id] !== undefined) {
      return;
    }

    if (this.roots[id] === undefined) {
      this.roots[id] = React.createRef();
    }

    if (this.state.embeddables[id] === undefined) {
      const embeddable = await this.props.container.getEmbeddable(id);
      const node = this.roots[id].current;
      if (this.mounted && node !== null && embeddable) {
        embeddable.renderInPanel(node, this.props.container);

        this.setState(prevState => ({
          loaded: { ...prevState.loaded, [id]: true },
          embeddables: {
            [id]: embeddable,
          },
        }));
      }
    }
  }
}
