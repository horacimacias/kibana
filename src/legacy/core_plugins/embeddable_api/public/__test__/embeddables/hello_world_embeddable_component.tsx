/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from 'react';
import { HelloWorldEmbeddable } from './hello_world_embeddable';

interface Props {
  helloWorldEmbeddable: HelloWorldEmbeddable;
}

interface State {
  fullName: string;
}

export class HelloWorldEmbeddableComponent extends React.Component<Props, State> {
  private unsubscribe?: () => void;
  private mounted: boolean = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      fullName: '',
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.unsubscribe = this.props.helloWorldEmbeddable.subscribeToOutputChanges(() => {
      if (this.mounted) {
        this.setState({ fullName: this.props.helloWorldEmbeddable.getOutput().name });
      }
    });
  }

  componentWillUnmount() {
    if (this.unsubscribe) {
      this.unsubscribe();
    }
    this.mounted = false;
  }

  render() {
    return (
      <div>
        Hello World!
        <br /> Sincerly, ${this.state.fullName}
      </div>
    );
  }
}
