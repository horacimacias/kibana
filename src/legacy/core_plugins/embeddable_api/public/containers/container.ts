/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import uuid from 'uuid';
import {
  Embeddable,
  EmbeddableFactoryRegistry,
  EmbeddableInput,
  EmbeddableOutput,
  ErrorEmbeddable,
  EmbeddableFactory,
} from '../embeddables';
import { ViewMode } from '../types';

export interface PanelState<E extends { [key: string]: unknown } = { [key: string]: unknown }> {
  savedObjectId?: string;

  embeddableId: string;
  // The type of embeddable in this panel. Will be used to find the factory in which to
  // load the embeddable.
  type: string;

  // Stores input for this embeddable that is specific to this embeddable. Other parts of embeddable input
  // will be derived from the container's input. **Any state in here will override any state derived from
  // the container.**
  partialInput: E; // TODO: rename explicitInput vs derivedInput that is taken from the container state.
}

export interface ContainerOutput extends EmbeddableOutput {
  embeddableLoaded: { [key: string]: boolean };
}

type NonNeverPropertyNames<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
type ObjectWithoutNever<T> = Pick<T, NonNeverPropertyNames<T>>;

export type EmbeddableInputMissingFromContainer<
  EI extends EmbeddableInput,
  CEI extends { [key: string]: unknown }
> = ObjectWithoutNever<{ [key in keyof EI]: CEI[key] extends EI[key] ? never : EI[key] }>;

export interface ContainerInput extends EmbeddableInput {
  hidePanelTitles?: boolean;
  panels: {
    [key: string]: PanelState;
  };
}

export abstract class Container<
  CEI extends Partial<EmbeddableInput> = {},
  EO extends EmbeddableOutput = EmbeddableOutput,
  I extends ContainerInput = ContainerInput,
  O extends ContainerOutput = ContainerOutput
> extends Embeddable<I, O> {
  public readonly isContainer: boolean = true;
  protected readonly embeddables: {
    [key: string]: Embeddable<EmbeddableInput, EO> | ErrorEmbeddable;
  } = {};
  public readonly embeddableFactories: EmbeddableFactoryRegistry;

  constructor(
    type: string,
    input: I,
    output: O,
    embeddableFactories: EmbeddableFactoryRegistry,
    parent?: Container
  ) {
    super(type, input, output, parent);
    this.embeddableFactories = embeddableFactories;
    this.initializeEmbeddables();
  }

  public getPartialInput(embeddableId: string) {
    return this.input.panels[embeddableId].partialInput;
  }

  public getViewMode() {
    return this.input.viewMode ? this.input.viewMode : ViewMode.EDIT;
  }

  public getHidePanelTitles() {
    return this.input.hidePanelTitles ? this.input.hidePanelTitles : false;
  }

  public updateInput(changes: Partial<I>): void {
    super.updateInput(changes);
  }

  public updateEmbeddableInput<EEI extends EmbeddableInput = EmbeddableInput>(
    id: string,
    changes: Partial<EEI>
  ) {
    if (!this.input.panels[id]) {
      throw new Error();
    }
    const panels = {
      panels: {
        ...this.input.panels,
        [id]: {
          ...this.input.panels[id],
          partialInput: {
            ...this.input.panels[id].partialInput,
            ...changes,
          },
        },
      },
    };
    this.updateInput(panels as Partial<I>);
  }

  public async addNewEmbeddable<
    EEI extends EmbeddableInput = EmbeddableInput,
    E extends Embeddable<EEI, EO> = Embeddable<EEI, EO>
  >(type: string, partialInput: Partial<EEI>): Promise<E | ErrorEmbeddable> {
    const factory = this.embeddableFactories.getFactoryByName(type) as EmbeddableFactory<
      EEI,
      EO,
      E
    >;
    const panelState = this.createNewPanelState<EEI>(factory, partialInput);
    this.setPanelState<EEI>(panelState);
    const embeddable = await factory.create<Container<CEI, EO, I, O>>(
      this.getInputForEmbeddable<EEI>(panelState.embeddableId),
      this
    );
    this.embeddables[embeddable.id] = embeddable;
    this.updateOutput({
      ...this.output,
      embeddableLoaded: {
        [panelState.embeddableId]: true,
      },
    });
    return embeddable;
  }

  private createNewExplicitEmbeddableInput<
    EEI extends EmbeddableInput = EmbeddableInput,
    E extends Embeddable<EEI, EO> = Embeddable<EEI, EO>
  >(id: string, factory: EmbeddableFactory<EEI, EO, E>, partial: Partial<EEI> = {}) {
    const inheritedInput = this.getInheritedInput(id) as { [key: string]: unknown };
    const defaults = factory.getDefaultInputParameters() as { [key: string]: unknown };

    // Container input overrides defaults.
    const explicitInput: { [key: string]: unknown } = partial;
    Object.keys(defaults).forEach(key => {
      if (inheritedInput[key] === undefined && explicitInput[key] === undefined) {
        explicitInput[key] = defaults[key];
      }
    });
    return explicitInput;
  }

  public async addSavedObjectEmbeddable<
    EEI extends EmbeddableInput = EmbeddableInput,
    E extends Embeddable<EEI, EO> = Embeddable<EEI, EO>
  >(type: string, savedObjectId: string): Promise<E | ErrorEmbeddable> {
    const factory = this.embeddableFactories.getFactoryByName(type) as EmbeddableFactory<
      EEI,
      EO,
      E
    >;
    const panelState = this.createNewPanelState<EEI>(factory);
    panelState.savedObjectId = savedObjectId;
    this.setPanelState<EEI>(panelState);
    const embeddable = await factory.createFromSavedObject(
      savedObjectId,
      this.getInputForEmbeddable<EEI>(panelState.embeddableId),
      this
    );
    this.embeddables[embeddable.id] = embeddable;

    this.updateOutput({
      ...this.output,
      embeddableLoaded: {
        [panelState.embeddableId]: true,
      },
    });
    return embeddable;
  }

  public removeEmbeddable(embeddableId: string) {
    const embeddable = this.getEmbeddable(embeddableId);
    embeddable.destroy();
    delete this.embeddables[embeddableId];

    const changedInput: { panels: { [key: string]: PanelState } } = {
      panels: {},
    };
    Object.values(this.input.panels).forEach(panel => {
      if (panel.embeddableId !== embeddableId) {
        changedInput.panels[panel.embeddableId] = panel;
      }
    });
    this.updateInput({ ...changedInput } as Partial<I>);
  }

  public getEmbeddable<E extends Embeddable<EmbeddableInput, EO> = Embeddable<EmbeddableInput, EO>>(
    id: string
  ): E {
    return this.embeddables[id] as E;
  }

  private setPanelState<EEI extends EmbeddableInput = EmbeddableInput>(
    panelState: PanelState<EmbeddableInputMissingFromContainer<EEI, CEI>>
  ) {
    this.updateInput({
      panels: {
        ...this.input.panels,
        [panelState.embeddableId]: panelState,
      },
    } as Partial<I>);
  }

  private updatePanelState<EEI extends EmbeddableInput = EmbeddableInput>(
    panelState: PanelState<EmbeddableInputMissingFromContainer<EEI, CEI>>
  ) {
    this.updateInput({
      panels: {
        ...this.input.panels,
        [panelState.embeddableId]: {
          ...this.input.panels[panelState.embeddableId],
          ...panelState,
        },
      },
    } as Partial<I>);
  }

  protected async loadEmbeddable<EEI extends EmbeddableInput = EmbeddableInput>(
    panelState: PanelState<EmbeddableInputMissingFromContainer<EEI, CEI>>
  ) {
    if (this.input.panels[panelState.embeddableId] === undefined) {
      throw new Error(`Panel with id ${panelState.embeddableId} does not exist in this container`);
    }

    const factory = this.embeddableFactories.getFactoryByName<EmbeddableFactory<EEI, EO>>(
      panelState.type
    );

    const embeddable = panelState.savedObjectId
      ? await factory.createFromSavedObject(
          panelState.savedObjectId,
          this.getInputForEmbeddable<EEI>(panelState.embeddableId),
          this
        )
      : await factory.create(this.getInputForEmbeddable<EEI>(panelState.embeddableId), this);
    this.embeddables[embeddable.id] = embeddable;
    this.updatePanelState<EEI>(panelState);

    this.updateOutput({
      embeddableLoaded: {
        [panelState.embeddableId]: true,
      },
    } as Partial<O>);
  }

  protected createNewPanelState<EEI extends EmbeddableInput = EmbeddableInput>(
    factory: EmbeddableFactory<EEI, EO>,
    partial: Partial<EEI> = {}
  ): PanelState<EmbeddableInputMissingFromContainer<EEI, CEI>> {
    const embeddableId = partial.id || uuid.v4();

    const explicitInput = this.createNewExplicitEmbeddableInput<EEI>(
      embeddableId,
      factory,
      partial
    ) as EmbeddableInputMissingFromContainer<EEI, CEI>;

    return {
      type: factory.name,
      embeddableId,
      partialInput: {
        id: embeddableId,
        ...explicitInput,
      },
    };
  }

  protected getPanelState<EI>(embeddableId: string) {
    if (this.input.panels[embeddableId] === undefined) {
      throw new Error(`No embeddable with id ${embeddableId}, this  ${JSON.stringify(this.input)}`);
    }
    const panelState: PanelState = this.input.panels[embeddableId];
    return panelState as PanelState<EI>;
  }

  protected abstract getInheritedInput(id: string): CEI;

  public getExplicitEmbeddableInput(id: string) {
    return this.input.panels[id].partialInput;
  }

  public getInputForEmbeddable<EEI extends EmbeddableInput = EmbeddableInput>(
    embeddableId: string
  ): EEI {
    const containerInput: CEI = this.getInheritedInput(embeddableId);
    const panelState = this.getPanelState<EmbeddableInputMissingFromContainer<EEI, CEI>>(
      embeddableId
    );

    const partialInput = panelState.partialInput as { [key: string]: unknown };
    const partialInputWithNoUndefinedParameters: { [key: string]: unknown } = {};

    Object.keys(panelState.partialInput).forEach(key => {
      if (partialInput[key] !== undefined) {
        partialInputWithNoUndefinedParameters[key] = partialInput[key];
      }
    });

    return ({
      ...containerInput,
      ...partialInputWithNoUndefinedParameters,
      // Typescript has difficulties with inferring this type but it is accurate with all
      // tests I tried. Could probably be revisted with future releases of TS to see if
      // it can accurately infer the type.
    } as unknown) as EEI;
  }

  private async initializeEmbeddables() {
    const promises = Object.values(this.input.panels).map(async panel => {
      await this.loadEmbeddable<EmbeddableInput>(panel as PanelState<
        EmbeddableInputMissingFromContainer<EmbeddableInput, CEI>
      >);
    });
    await Promise.all(promises);
  }
}
