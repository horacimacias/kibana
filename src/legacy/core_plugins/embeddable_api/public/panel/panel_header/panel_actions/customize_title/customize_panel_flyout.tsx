/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { Component } from 'react';

import { EuiButton, EuiFlyoutBody, EuiFlyoutHeader, EuiTitle } from '@elastic/eui';
import { Embeddable } from '../../../../';
import { CustomizeTitleForm } from './customize_title_form';

interface CustomizePanelProps {
  embeddable: Embeddable;
  updateTitle: (newTitle: string | undefined) => void;
}

interface State {
  title: string | undefined;
}

export class CustomizePanelFlyout extends Component<CustomizePanelProps, State> {
  constructor(props: CustomizePanelProps) {
    super(props);
    this.state = {
      title: props.embeddable.getTitle(),
    };
  }

  updateTitle = (title: string | undefined) => {
    this.setState({ title });
  };

  reset = () => {
    this.props.updateTitle(undefined);
  };

  public render() {
    return (
      <React.Fragment>
        <EuiFlyoutHeader>
          <EuiTitle size="s" data-test-subj="customizePanelTitle">
            <h1>{this.props.embeddable.getTitle()}</h1>
          </EuiTitle>
        </EuiFlyoutHeader>
        <EuiFlyoutBody>
          <CustomizeTitleForm
            title={this.state.title || this.props.embeddable.getTitle()}
            onReset={this.reset}
            onUpdatePanelTitle={this.updateTitle}
          />
          <EuiButton onClick={() => this.props.updateTitle(this.state.title)}>
            Save & Close
          </EuiButton>
        </EuiFlyoutBody>
      </React.Fragment>
    );
  }
}
