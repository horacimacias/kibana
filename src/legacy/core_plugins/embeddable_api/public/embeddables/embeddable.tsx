/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { I18nProvider } from '@kbn/i18n/react';
import { isEqual } from 'lodash';
import React, { ReactNode } from 'react';
import ReactDOM from 'react-dom';
import { Adapters } from 'ui/inspector';
import { BehaviorSubject } from 'rxjs';
import * as Rx from 'rxjs';
import { Container } from '../containers';
import { EmbeddablePanel } from '../panel';
import { Trigger } from '../triggers';
import { OutputSpec, ViewMode } from '../types';
import { IEmbeddable } from './i_embeddable';

export interface EmbeddableInput {
  viewMode?: ViewMode;
  title?: string;
  id: string;
  savedObjectId?: string;
}

export interface EmbeddableOutput {
  editUrl?: string;
  title?: string;
  editable?: boolean;
}

export class Embeddable<
  I extends EmbeddableInput = EmbeddableInput,
  O extends EmbeddableOutput = EmbeddableOutput
> implements IEmbeddable<I, O> {
  public readonly parent?: Container;
  public readonly isContainer: boolean = false;
  public readonly type: string;
  public readonly id: string;

  private readonly input$: Rx.BehaviorSubject<I>;
  private readonly output$: Rx.BehaviorSubject<O>;

  // protected anyChangeListeners: Array<
  //   <EI extends I, EO extends O>(changes: { output?: Partial<EO>; input?: Partial<EI> }) => void
  // > = [];
  // protected inputChangeListeners: Array<(<EI extends I>(input: Partial<EI>) => void)> = [];
  // protected outputChangeListeners: Array<(<EO extends O>(output: Partial<EO>) => void)> = [];
  protected output: O;
  protected input: I;

  private panelContainer?: Element;
  private parentSubscription?: Rx.Subscription;
  private destoyed: boolean = false;

  constructor(type: string, input: I, output: O, parent?: Container) {
    this.type = type;
    this.id = input.id;
    this.output = output;
    this.input = input;
    this.parent = parent;

    if (!input.id) {
      throw new Error('Id required');
    }

    if (parent) {
      this.parentSubscription = Rx.merge(parent.getInput$(), parent.getOutput$()).subscribe(() => {
        const newInput = parent.getInputForEmbeddable<I>(this.id);
        this.onParentInputChanged(newInput);
      });

      // this.parentSubscription = parent.subscribeToChanges(() => {
      //   const newInput = parent.getInputForEmbeddable<I>(this.id);
      //   this.onParentInputChanged(newInput);
      // });
    }

    this.input$ = new Rx.BehaviorSubject<I>(this.input);
    this.output$ = new Rx.BehaviorSubject<O>(this.output);
  }

  public getInput$(): Readonly<Rx.Observable<EmbeddableInput>> {
    return this.input$.asObservable();
  }

  public getOutput$(): Readonly<Rx.Observable<EmbeddableOutput>> {
    return this.output$.asObservable();
  }

  public getOutput(): Readonly<O> {
    return this.output;
  }

  public getInput(): Readonly<I> {
    return this.input;
  }

  public getTitle() {
    return this.input.title || this.output.title;
  }

  public updateInput(changes: Partial<I>): void {
    if (this.destoyed) {
      throw new Error('Embeddable has been destroyed');
    }
    if (this.parent) {
      // Ensures state changes flow from container downward.
      this.parent.updateEmbeddableInput<I>(this.id, changes);
    } else {
      const newInput = {
        ...this.input,
        ...changes,
      };
      if (!isEqual(this.input, newInput)) {
        this.input = newInput;
        this.input$.next(this.input);
        // this.emitInputChanged(changes);
      }
    }
  }

  public getOutputSpec(trigger?: Trigger): OutputSpec {
    return {};
  }

  // public subscribeToInputChanges(listener: (input: Partial<I>) => void) {
  //   if (this.destoyed) {
  //     throw new Error('Embeddable has been destroyed');
  //   }
  //   this.inputChangeListeners.push(listener);
  //   const unsubscribe = () => {
  //     this.inputChangeListeners.splice(this.inputChangeListeners.indexOf(listener), 1);
  //   };
  //   return unsubscribe;
  // }

  // public subscribeToOutputChanges(listener: (output: Partial<O>) => void) {
  //   if (this.destoyed) {
  //     throw new Error('Embeddable has been destroyed');
  //   }
  //   this.outputChangeListeners.push(listener);
  //   const unsubscribe = () => {
  //     this.outputChangeListeners.splice(this.outputChangeListeners.indexOf(listener), 1);
  //   };
  //   return unsubscribe;
  // }

  // public subscribeToChanges(
  //   listener: (changes: { input?: Partial<I>; output?: Partial<O> }) => void
  // ) {
  //   if (this.destoyed) {
  //     throw new Error('Embeddable has been destroyed');
  //   }
  //   this.anyChangeListeners.push(listener);
  //   const unsubscribe = () => {
  //     this.anyChangeListeners.splice(this.anyChangeListeners.indexOf(listener), 1);
  //   };
  //   return unsubscribe;
  // }

  public supportsTrigger(trigger: Trigger) {
    return false;
  }

  public renderInPanel(node: HTMLElement | Element, container?: Container) {
    if (this.destoyed) {
      throw new Error('Embeddable has been destroyed');
    }
    this.panelContainer = node;

    ReactDOM.render(
      <I18nProvider>
        <EmbeddablePanel embeddable={this} />
      </I18nProvider>,
      node
    );
  }

  public render(domNode: HTMLElement | ReactNode): void {
    if (this.destoyed) {
      throw new Error('Embeddable has been destroyed');
    }
    return;
  }

  /**
   * An embeddable can return inspector adapters if it want the inspector to be
   * available via the context menu of that panel.
   * @return Inspector adapters that will be used to open an inspector for.
   */
  public getInspectorAdapters(): Adapters | undefined {
    return undefined;
  }

  public destroy(): void {
    this.destoyed = true;
    if (this.panelContainer) {
      ReactDOM.unmountComponentAtNode(this.panelContainer);
    }
    if (this.parentSubscription) {
      this.parentSubscription.unsubscribe();
    }
    return;
  }

  // protected emitOutputChanged(changes: Partial<O>) {
  //   // Create copies to avoid issues if listeners are unsubscribed by nested changes propagatings.
  //   [...this.outputChangeListeners].forEach(listener => listener(changes));
  //   [...this.anyChangeListeners].forEach(listener => listener({ output: changes }));
  // }

  protected updateOutput(outputChanges: Partial<O>): void {
    const newOutput = {
      ...this.output,
      ...outputChanges,
    };
    if (!isEqual(this.output, newOutput)) {
      this.output = newOutput;
      this.output$.next(this.output);
    }
  }

  private onParentInputChanged(newInput: I) {
    if (!isEqual(this.input, newInput)) {
      this.input = { ...newInput };

      this.input$.next(this.input);
    }
  }

  // private emitInputChanged(changes: Partial<I>) {
  //   this.input$.next(this.input);
  // }
}
