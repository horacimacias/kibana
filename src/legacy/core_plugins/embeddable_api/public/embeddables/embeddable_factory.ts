/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { SavedObjectMetaData } from 'ui/saved_objects/components/saved_object_finder';
import { SavedObjectAttributes } from '../../../../server/saved_objects';
import { Embeddable, EmbeddableInput, EmbeddableOutput } from './embeddable';
import { ErrorEmbeddable } from './error_embeddable';
import { Container } from '../containers';

export interface EmbeddableInstanceConfiguration {
  id: string;
  savedObjectId?: string;
}

export type AnyEmbeddableFactory = EmbeddableFactory;

export interface PropertySpec {
  displayName: string;
  accessPath: string;
  id: string;
  description: string;
  value?: string;
}

export interface OutputSpec {
  [key: string]: PropertySpec;
}

/**
 * The EmbeddableFactory creates and initializes an embeddable instance
 */
export abstract class EmbeddableFactory<
  I extends EmbeddableInput = EmbeddableInput,
  O extends EmbeddableOutput = EmbeddableOutput,
  E extends Embeddable<I, O> = Embeddable<I, O>,
  T extends SavedObjectAttributes = SavedObjectAttributes
> {
  public readonly name: string;
  public readonly savedObjectMetaData?: SavedObjectMetaData<T>;

  /**
   *
   * @param name - a unique identified for this factory, which will be used to map an embeddable spec to
   * a factory that can generate an instance of it.
   */
  constructor({
    name,
    savedObjectMetaData,
  }: {
    name: string;
    savedObjectMetaData?: SavedObjectMetaData<T>;
  }) {
    this.name = name;
    this.savedObjectMetaData = savedObjectMetaData;
  }

  public abstract getOutputSpec(): OutputSpec;

  public isEditable() {
    return true;
  }

  public getDefaultInputParameters(): Partial<I> {
    return {};
  }

  /**
   */
  public createFromSavedObject(
    savedObjectId: string,
    partialInput: Partial<I>,
    parent?: Container
  ): Promise<E | ErrorEmbeddable> {
    throw new Error(`Creation from saved object not supported by type ${this.name}`);
  }

  /**
   */
  public abstract create<C extends Container = Container>(
    initialInput: I,
    parent?: C
  ): Promise<E | ErrorEmbeddable>;
}
