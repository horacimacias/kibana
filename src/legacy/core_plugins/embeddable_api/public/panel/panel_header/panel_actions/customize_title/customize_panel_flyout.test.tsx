/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

jest.mock('ui/metadata', () => ({
  metadata: {
    branch: 'my-metadata-branch',
    version: 'my-metadata-version',
  },
}));

jest.mock('ui/capabilities', () => ({
  uiCapabilities: {
    visualize: {
      save: true,
    },
  },
}));

import React from 'react';
import {
  HELLO_WORLD_EMBEDDABLE,
  HelloWorldEmbeddableFactory,
  HelloWorldContainer,
  HelloWorldEmbeddable,
  HelloWorldInput,
} from '../../../../__test__/index';

import { CustomizePanelFlyout } from './customize_panel_flyout';
import { Container } from 'plugins/embeddable_api/containers';
import { EmbeddableFactoryRegistry, isErrorEmbeddable } from 'plugins/embeddable_api/embeddables';
import { shallowWithIntl } from 'test_utils/enzyme_helpers';

let container: Container;
let embeddable: HelloWorldEmbeddable;

beforeEach(async () => {
  const embeddableFactories = new EmbeddableFactoryRegistry();
  embeddableFactories.registerFactory(new HelloWorldEmbeddableFactory());
  container = new HelloWorldContainer({ id: '123', panels: {} }, embeddableFactories);
  const helloEmbeddable = await container.addNewEmbeddable<HelloWorldInput, HelloWorldEmbeddable>(
    HELLO_WORLD_EMBEDDABLE,
    {
      firstName: 'Joe',
    }
  );
  if (isErrorEmbeddable(helloEmbeddable)) {
    throw new Error('Error creating new hello world embeddable');
  } else {
    embeddable = helloEmbeddable;
  }
});

test('matches snapshot', async () => {
  const component = shallowWithIntl(
    <CustomizePanelFlyout embeddable={embeddable} updateTitle={() => {}} />
  );

  expect(component).toMatchSnapshot();
});
