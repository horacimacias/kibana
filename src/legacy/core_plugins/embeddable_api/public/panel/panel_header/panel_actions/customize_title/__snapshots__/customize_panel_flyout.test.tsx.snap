// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Fragment>
  <EuiFlyoutHeader>
    <EuiTitle
      data-test-subj="customizePanelTitle"
      size="s"
    >
      <h1>
        Joe foo
      </h1>
    </EuiTitle>
  </EuiFlyoutHeader>
  <EuiFlyoutBody>
    <InjectIntl(CustomizeTitleFormUi)
      onReset={[Function]}
      onUpdatePanelTitle={[Function]}
      title="Joe foo"
    />
    <EuiButton
      color="primary"
      fill={false}
      iconSide="left"
      onClick={[Function]}
      size="m"
      type="button"
    >
      Save & Close
    </EuiButton>
  </EuiFlyoutBody>
</Fragment>
`;
